src/error/error_impl.rs

```diff
@@ -49,8 +48,23 @@ impl ErrorImpl {
         });
         match ret {
             Some(ret) => {
-                // forget self (backtrace is dropped, failure is moved
-                mem::forget(self);
+                // deallocate the box without dropping the inner parts
+                #[cfg(has_global_alloc)] {
+                    use std::alloc::{dealloc, Layout};
+                    unsafe {
+                        let layout = Layout::for_value(&*self.inner);
+                        let ptr = Box::into_raw(self.inner);
+                        dealloc(ptr as *mut u8, layout);
+                    }
+                }
+
+                // slightly leaky versions of the above thing which makes the box
+                // itself leak.  There is no good way around this as far as I know.
+                #[cfg(not(has_global_alloc))] {
+                    use core::mem;
+                    mem::forget(self);
```

Comment:
This is pure curiosity, but do you mind explaining why this leaks memory?

If I understand correctly, the `ptr::read(fail as *const T)` does a memcopy `T` and moves `T` out of its box without deallocating the box. But when we to `mem::forget(self)` why isn't the box's memory not deallocated?

Reply:
`mem::forget` tells it not to run a destructor. By moving out the fields all inner fields were destroyed but the heap allocation for the containing struct in the box is forgotten and leaks.